### PR TITLE
test: add edge-case tests for datetime_to_epoch function

### PR DESCRIPTION
**Pull request by KrupaVadher**
## Summary
Add comprehensive edge-case tests for the `datetime_to_epoch()` utility function to improve test coverage and reliability.

## Changes
- Test epoch boundary conditions (1970-01-01 00:00:00)
- Test timezone-aware datetime conversion
- Test DST transition handling
- Test microsecond precision
- Test far future date handling

## Testing
All tests pass locally and cover critical edge cases for date/time conversion in Superset's data visualization pipeline.

### CODE REVIEW

**Comment by sadpandajoe** [comment]
Superset uses Git pre-commit hooks courtesy of [pre-commit](https://pre-commit.com/). To install run the following:

```
pip3 install -r requirements/development.txt
pre-commit install
```

A series of checks will now run when you make a git commit.

Alternatively it is possible to run pre-commit by running pre-commit manually:

```
pre-commit run --all-files
```

**Comment by rusackas** [comment]
Looks like this needs some assessment of the bot review comments, if you don't mind taking a look at them. Appreciate the PR!

**Review by rusackas — APPROVED** [review]
Thanks @KrupaVadher, the extra coverage on the epoch boundary and the tz cases is welcome, and I went through the expected values... they all check out. LGTM!

One tiny thing, not blocking: the 'DST' test uses two unambiguous times, so it's really just a plain UTC subtraction and the `is_dst` flags are actually backwards. Flipping them and tightening the tolerance to an exact `==` would make it mean what the name says. Happy to merge either way.

File: tests/unit_tests/utils/test_date_parsing.py [hunk 1]
```diff
@@ -254,3 +257,91 @@ def test_warning_suppression():
...
+    result = datetime_to_epoch(epoch_dt)
+    assert result == 0.0, f"Epoch datetime should be 0.0, got {result}"
+
+
+def test_datetime_to_epoch_naive_one_second_after():
+    """Test naive datetime 1 second after epoch"""
+    dt = datetime(1970, 1, 1, 0, 0, 1)
+    result = datetime_to_epoch(dt)
+    expected = 1000.0  # 1 second * 1000 ms
+    assert result == expected, f"Expected {expected}ms, got {result}ms"
```
Comment (rusackas) [inline tests/unit_tests/utils/test_date_parsing.py:280]: Good catch, agreed. I can't push directly since maintainer edits are disabled on this PR, but here's the addition as a one-click suggestion:

```suggestion
    assert result == expected, f"Expected {expected}ms, got {result}ms"


def test_datetime_to_epoch_naive_one_second_before():
    """Test naive datetime 1 second before epoch returns a negative value"""
    dt = datetime(1969, 12, 31, 23, 59, 59)
    result = datetime_to_epoch(dt)
    expected = -1000.0  # 1 second before epoch, in ms
    assert result == expected, f"Expected {expected}ms, got {result}ms"
```

File: tests/unit_tests/utils/test_date_parsing.py [hunk 2]
```diff
@@ -254,3 +254,89 @@ def test_warning_suppression():
...
+    epoch_dt = datetime(1970, 1, 1, 0, 0, 0)
+    result = datetime_to_epoch(epoch_dt)
+    assert result == 0.0, f"Epoch datetime should be 0.0, got {result}"
+
+
+def test_datetime_to_epoch_naive_one_second_after():
+    """Test naive datetime 1 second after epoch"""
+    dt = datetime(1970, 1, 1, 0, 0, 1)
+    result = datetime_to_epoch(dt)
+    expected = 1000.0  # 1 second * 1000 ms
```
[the remark this answers is not in the document]
Comment (rusackas) [reply tests/unit_tests/utils/test_date_parsing.py:279]: Agreed on the coverage gap — since maintainer edits are disabled on this PR I posted the addition as a suggestion below (comment #3565299507) rather than pushing directly.

File: tests/unit_tests/utils/test_date_parsing.py [hunk 3]
```diff
@@ -254,3 +257,91 @@ def test_warning_suppression():
...
+    # Use a known DST transition date in US/Eastern
+    # 2023-03-12: Spring forward (2 AM becomes 3 AM, gap of 1 hour)
+    eastern = pytz.timezone("US/Eastern")
+
+    # Create datetime before DST transition
+    dt_before_dst = eastern.localize(datetime(2023, 3, 12, 1, 59, 59), is_dst=True)
+    result_before = datetime_to_epoch(dt_before_dst)
+
+    # Create datetime after DST transition
+    dt_after_dst = eastern.localize(datetime(2023, 3, 12, 3, 0, 1), is_dst=False)
```
[the remark this answers is not in the document]
Comment (rusackas) [reply tests/unit_tests/utils/test_date_parsing.py]: Already addressed in a later commit (6a095a04e6) — the flags were flipped (`is_dst=False` before the transition, `is_dst=True` after) and the tolerance tightened to an exact `==`.

File: tests/unit_tests/utils/test_date_parsing.py [hunk 4]
```diff
@@ -254,3 +257,91 @@ def test_warning_suppression():
...
     assert len(warnings_list) == 0  # Should suppress all format inference warnings
     assert pd.api.types.is_datetime64_any_dtype(df["date"])  # Should still parse dates
+
+
+# ============================================================================
+# NEW TESTS FOR datetime_to_epoch() - Edge case coverage
+# ============================================================================
+
+
+def test_datetime_to_epoch_naive_at_epoch():
```
[the remark this answers is not in the document]
Comment (rusackas) [reply tests/unit_tests/utils/test_date_parsing.py:267]: Not applying — none of the other test functions in this file (existing or new) have return type hints either, so this would be inconsistent with the file's established convention. Would be a separate, file-wide cleanup, not something to single out on this one function.

File: tests/unit_tests/utils/test_date_parsing.py [hunk 5]
```diff
@@ -254,3 +254,89 @@ def test_warning_suppression():
...
+    # Create datetime before DST transition
+    dt_before_dst = eastern.localize(datetime(2023, 3, 12, 1, 59, 59), is_dst=True)
+    result_before = datetime_to_epoch(dt_before_dst)
+
+    # Create datetime after DST transition
+    dt_after_dst = eastern.localize(datetime(2023, 3, 12, 3, 0, 1), is_dst=False)
+    result_after = datetime_to_epoch(dt_after_dst)
+
+    # The difference should be only 2 seconds, not 1 hour + 2 seconds
+    # (because of the DST jump, 1:59:59 EST -> 3:00:01 EDT)
```
[the remark this answers is not in the document]
Comment (rusackas) [reply tests/unit_tests/utils/test_date_parsing.py]: Not applying — `datetime_to_epoch()` doesn't do its own DST arithmetic; it just calls `.astimezone(pytz.utc)` on whatever aware datetime it's given, so all DST/offset resolution happens inside `pytz` when the datetime is constructed via `localize()`, not in this function. Testing a literal nonexistent local time (e.g. 2:30 AM on the spring-forward day) would exercise pytz's own ambiguous-time handling, not any logic in `datetime_to_epoch` itself. The existing test already covers the meaningful assertion — that the UTC-ms difference across a DST-crossing boundary is computed correctly.
